rfcs/0157-more-granular-control.md

```diff
@@ -0,0 +1,241 @@
+---
+feature: more-granular-control
+start-date: 2023-08-02
+author: Animesh Sahu
+co-authors: (find a buddy later to help out with the RFC)
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+To grant more granular low-level control over the system. Majorly:
+
+* `nix-cli`: Introduce direct access to the store and binary cache.
+  - This improvement will facilitate easier binary cache setup, upstream caching, and shallow (non-closure) copies.
+  - Users will be able to pull specific paths directly from the cache without relying on derivation realisation as an abstraction.
+
+* `nixos`: Decouple various low-level components of the operating system and make everything "opt-in."
+  - This change aims to reduce the disparity observed when using nix inside NixOS and outside NixOS.
+  - By making the infrastructure flexible and allowing components to be explicitly chosen, users can adapt to future breaking changes, such as the introduction of pipewire and wayland, even for init systems and bootloaders.
+
+* `flakes`: Enforce the use of flakes for scripting operations.
+  - Currently, script infrastructure often involves a `script -> nix language -> script` abstraction, which can make it challenging to modify, reuse, or apply fixes to the underlying script. Examples of this include nixos-rebuild, nixos-generate, and home-manager.
+  - With the enforcement of flakes, the former `script ->` abstraction will be reduced, and users can directly use a `nix` command to build the script. The activation script `./result/bin/activation-script` can then be executed to complete the task.
```

Comment:
I'm not really comfortable with oldschool syntax, I'll carefully read them up tomorrow and see if prime purpose - indirection removal be done for derivations that give rise to yet another shell script. If it can be served using that, I'll see how RFC has to be modified that way.